File: rb/companydb.rb
require_relative 'company'
require_relative 'employee'
require_relative 'customer'

class CompanyDB
  def initialize(company_name)
    @company = Company.new
    @company_name = company_name
    load_data
  end
  
  #method that checkes if file is present, if so it loads company data, if not it creates new file
  def load_data
    filename = "#{@company_name.downcase}.dat"
    if File.exist?(filename)
      File.open(filename, 'r') do |file|
        num_employees = file.readline.to_i
        num_employees.times do
          name = file.readline.chomp
          email = file.readline.chomp
          phone = file.readline.chomp
          salary = file.readline.chomp.to_f
          employee = Employee.new(name, email, phone, salary)
          @company.add_employee(employee)
        end
  
        num_customers = file.readline.to_i
        num_customers.times do
          customer_name = file.readline.chomp
          customer_email = file.readline.chomp
          customer_phone = file.readline.chomp
          num_purchases = file.readline.to_i
          customer = Customer.new(customer_name, customer_email, customer_phone)
  
          num_purchases.times do
            item = file.readline.chomp
            quantity = file.readline.to_i
            cost = file.readline.to_f
            purchase = Purchase.new(item, quantity, cost)
            customer.purchases << purchase
          end
  
          @company.customers << customer
        end
      end
    else
      # If the file doesn't exist, create a new file with the company name
      File.open(filename, 'w') {}
    end
  end

  # method to save company data in the proper format to a respective .dat file 
  def save_data
    filename = "#{@company_name.downcase}.dat"
    File.open(filename, 'w') do |file|
      file.puts(@company.employees.size)
      @company.employees.each do |employee|
        file.puts(employee.name)
        file.puts(employee.email)
        file.puts(employee.phone)
        file.puts(employee.salary)
      end
  
      file.puts(@company.customers.size)
      @company.customers.each do |customer|
        file.puts(customer.name)
        file.puts(customer.email)
        file.puts(customer.phone)
        file.puts(customer.purchases.size)
        customer.purchases.each do |purchase|
          file.puts(purchase.item)
          file.puts(purchase.quantity)
          file.puts(purchase.cost)
        end
      end
    end
  end

  def list_employees
    @company.list_employees
  end

  # Method handles the main menu of the program
  def main_menu
    loop do
      puts "    MAIN MENU"
      puts "1.) Employees"
      puts "2.) Sales"
      puts "3.) Quit"
      print "\nChoice? "
      choice = gets.chomp.to_i

      case choice
      when 1
        employees_menu
      when 2
        sales_menu
      when 3
        save_data
        exit
      else
        puts "Invalid choice. Please try again."
      end
    end
  end

  #Method handles the Employees menu (1)
  def employees_menu
    loop do
      # List employees to match sample
      @company.list_employees

      print "(A)dd Employee or (M)ain Menu? "
      choice = gets.chomp.upcase

      case choice
      when 'A'
        add_employee
      when 'M'
        break
      else
        puts "Invalid choice. Please try again."
      end
    end
  end

  #add employee to company
  def add_employee
    print "Name: "
    name = gets.chomp
    print "Email: "
    email = gets.chomp
    print "Phone: "
    phone = gets.chomp
    print "Salary: "
    salary = gets.chomp.to_f

    employee = Employee.new(name, email, phone, salary)
    @company.add_employee(employee)  # Add the employee to the company
    #puts "#{employee.name} <#{employee.email}>  Phone: #{employee.phone} Salary: $#{'%.2f' % employee.salary}" #debugging
  end

  # Method hanldes the Sales menu (2)
  def sales_menu
    loop do
      print "(A)dd Customer, Enter a (S)ale, (V)iew Customer, or (M)ain Menu? "
      choice = gets.chomp.upcase

      case choice
      when 'A'
        add_customer
      when 'S'
        enter_sale
      when 'V'
        view_customer
      when 'M'
        break
      else
        puts "Invalid choice. Please try again."
      end
    end
  end

  # method handles adds customer choice
  def add_customer
    print "Name: "
    name = gets.chomp
    print "Email: "
    email = gets.chomp
    print "Phone: "
    phone = gets.chomp

    customer = Customer.new(name, email, phone)
    @company.customers << customer
  end

  #Method handles the sale choice
  def enter_sale
    if @company.customers.empty?
      puts "Error: No Customers." #promtps an error message if customers is empty
      return
    end

    puts "Customers:"
    @company.customers.each_with_index do |customer, index|
      puts "#{index + 1}.) #{customer.name}"
    end

    print "Choice? "
    customer_choice = gets.chomp.to_i - 1

    # Adds purchase to selected customer 
    if customer_choice >= 0 && customer_choice < @company.customers.length
      customer = @company.customers[customer_choice]

      print "Item: "
      item = gets.chomp
      print "Quantity: "
      quantity = gets.chomp.to_i
      print "Cost: "
      cost = gets.chomp.to_f

      purchase = Purchase.new(item, quantity, cost)
      customer.purchases << purchase
    else
      puts "Invalid choice. Sale not entered."
    end
  end

  # Handles view choice
  def view_customer
    if @company.customers.empty?
      puts "Error: No Customers."
      return
    end

    puts "Customers:"
    @company.customers.each_with_index do |customer, index|
      puts "#{index + 1}.) #{customer.name}"
    end

    print "Choice? "
    customer_choice = gets.chomp.to_i - 1

    if customer_choice >= 0 && customer_choice < @company.customers.length
      customer = @company.customers[customer_choice]

      # Prints Purchase history of selected customer 
      puts "#{customer.name} <#{customer.email}>  Phone: #{customer.phone}\n\nOrder History"
      puts "Item                   Price  Quantity   Total"
      customer.purchases.each do |purchase|
        total = purchase.quantity * purchase.cost
        puts "#{purchase.item.ljust(22)}#{purchase.cost.to_s.ljust(7)}#{purchase.quantity.to_s.ljust(10)}#{total}"
      end
    else
      puts "Invalid choice. Cannot view customer."
    end
  end
end

# Main program
print "Company Name: "
company_name = gets.chomp
company_db = CompanyDB.new(company_name)
company_db.main_menu

File: Lua/Lua/companydb.lua
-- companydb Class that is entry point of program
-- Contains program menus and file handling

-- Import statements for classes
local Employee = require("employee")
local Customer = require("customer")
local Purchase = require("purchase")
local Company = require("company")

CompanyDB = {}
CompanyDB.__index = CompanyDB

function CompanyDB.new(company_name)
    local self = setmetatable({}, CompanyDB)
    self.company = Company.new()
    self.company_name = company_name
    self:load_data()
    return self
end

-- Loads employees, customers, and sales based on file formatting
function CompanyDB:load_data()
    local filename = string.lower(self.company_name) .. ".dat"
    local file = io.open(filename, "r")
    if file then
        local content = file:read("*a")
        file:close()

        local lines = {}
        for line in content:gmatch("[^\r\n]+") do
            table.insert(lines, line)
        end

        local numEmployees = tonumber(lines[1])
        local currentLine = 2
        for i = 1, numEmployees do
            local name = lines[currentLine]
            local email = lines[currentLine + 1]
            local phone = lines[currentLine + 2]
            local salary = tonumber(lines[currentLine + 3])
            local employee = Employee.new(name, email, phone, salary)
            self.company:add_employee(employee)
            currentLine = currentLine + 4
        end

        local numCustomers = tonumber(lines[currentLine])
        currentLine = currentLine + 1
        for i = 1, numCustomers do
            local customerName = lines[currentLine]
            local customerEmail = lines[currentLine + 1]
            local customerPhone = lines[currentLine + 2]
            local numPurchases = tonumber(lines[currentLine + 3])
            local customer = Customer.new(customerName, customerEmail, customerPhone)
            currentLine = currentLine + 4

            for j = 1, numPurchases do
                local item = lines[currentLine]
                local quantity = tonumber(lines[currentLine + 1])
                local cost = tonumber(lines[currentLine + 2])
                local purchase = Purchase.new(item, quantity, cost)
                customer.purchases[#customer.purchases + 1] = purchase
                currentLine = currentLine + 3
            end

            self.company.customers[#self.company.customers + 1] = customer
        end

        print("Data loaded successfully.")
    else
        print("Error: Unable to load data. Creating a new company database.")
        -- Create a new file if it doesn't exist
        local newFile = io.open(filename, "w")
        newFile:close()
    end
end

function CompanyDB:save_data()
    local filename = string.lower(self.company_name) .. ".dat"
    local file = io.open(filename, "w")
    if file then
        -- Write employees data
        file:write(#self.company.employees .. "\n")
        for _, employee in ipairs(self.company.employees) do
            file:write(employee.name .. "\n")
            file:write(employee.email .. "\n")
            file:write(employee.phone .. "\n")
            file:write(employee.salary .. "\n")
        end

        -- Write customers data
        file:write(#self.company.customers .. "\n")
        for _, customer in ipairs(self.company.customers) do
            file:write(customer.name .. "\n")
            file:write(customer.email .. "\n")
            file:write(customer.phone .. "\n")
            file:write(#customer.purchases .. "\n")
            for _, purchase in ipairs(customer.purchases) do
                file:write(purchase.item .. "\n")
                file:write(purchase.quantity .. "\n")
                file:write(purchase.cost .. "\n")
            end
        end

        file:close()
        print("Data saved successfully.")
    else
        print("Error: Unable to save data.")
    end
end

function CompanyDB:list_employees()
    self.company:list_employees()
end

function CompanyDB:employees_menu()
    while true do
        -- List employees
        print("Employees:")
        self.company:list_employees()

        io.write("(A)dd Employee or (M)ain Menu? ")
        local choice = io.read():upper()

        if choice == "A" then
            self:add_employee()
        elseif choice == "M" then
            break
        else
            print("Invalid choice. Please try again.")
        end
    end
end

function CompanyDB:add_employee()
    io.write("Enter employee details:\nName: ")
    local name = io.read()
    io.write("Email: ")
    local email = io.read()
    io.write("Phone: ")
    local phone = io.read()
    io.write("Salary: ")
    local salary = tonumber(io.read())

    local employee = Employee.new(name, email, phone, salary)
    self.company:add_employee(employee)
    print("Employee added successfully.")
end

function CompanyDB:sales_menu()
    while true do
        io.write("(A)dd Customer, Enter a (S)ale, (V)iew Customer, or (M)ain Menu? ")
        local choice = io.read():upper()

        if choice == "A" then
            self:add_customer()
        elseif choice == "S" then
            self:enter_sale()
        elseif choice == "V" then
            self:view_customer()
        elseif choice == "M" then
            break
        else
            print("Invalid choice. Please try again.")
        end
    end
end

function CompanyDB:add_customer()
    io.write("Enter customer details:\nName: ")
    local name = io.read()
    io.write("Email: ")
    local email = io.read()
    io.write("Phone: ")
    local phone = io.read()

    local customer = Customer.new(name, email, phone)
    self.company.customers[#self.company.customers + 1] = customer
    print("Customer added successfully.")
end

function CompanyDB:enter_sale()
    if #self.company.customers == 0 then
        print("Error: No Customers.")
        return
    end

    print("Customers:")
    for i, customer in ipairs(self.company.customers) do
        print(i .. ".) " .. customer.name)
    end

    io.write("Enter customer number: ")
    local customer_choice = tonumber(io.read())
    if customer_choice and customer_choice >= 1 and customer_choice <= #self.company.customers then
        local customer = self.company.customers[customer_choice]
        io.write("Item: ")
        local item = io.read()
        io.write("Quantity: ")
        local quantity = tonumber(io.read())
        io.write("Cost: ")
        local cost = tonumber(io.read())

        local purchase = Purchase.new(item, quantity, cost)
        customer.purchases[#customer.purchases + 1] = purchase
        print("Sale entered successfully.")
    else
        print("Invalid choice. Sale not entered.")
    end
end

function CompanyDB:view_customer()
    if #self.company.customers == 0 then
        print("Error: No Customers.")
        return
    end

    print("Customers:")
    for i, customer in ipairs(self.company.customers) do
        print(i .. ".) " .. customer.name)
    end

    io.write("Enter customer number: ")
    local customer_choice = tonumber(io.read())
    if customer_choice and customer_choice >= 1 and customer_choice <= #self.company.customers then
        local customer = self.company.customers[customer_choice]
        print(customer.name .. " <" .. customer.email .. ">  Phone: " .. customer.phone .. "\n\nOrder History")
        print("Item                   Price   Quantity   Total")

        -- handles padding in sales view to neatly display price/qty/total
        for _, purchase in ipairs(customer.purchases) do 
            local itemName = purchase.item .. string.rep(" ", 23 - #purchase.item) 
            local price = string.format("%.2f", purchase.cost) .. string.rep(" ", 9 - #tostring(purchase.cost))
            local quantity = tostring(purchase.quantity) .. string.rep(" ", 11 - #tostring(purchase.quantity))
            local total = string.format("%.2f", purchase.quantity * purchase.cost)
            print(itemName .. price .. quantity .. total)
        end
    else
        print("Invalid choice. Cannot view customer.")
    end
end

function CompanyDB:main_menu()
    while true do
        print("    MAIN MENU")
        print("1.) Employees")
        print("2.) Sales")
        print("3.) Quit")
        io.write("Choice? ")
        local choice = tonumber(io.read())

        if choice == 1 then
            self:employees_menu()
        elseif choice == 2 then
            self:sales_menu()
        elseif choice == 3 then
            self:save_data()
            break
        else
            print("Invalid choice. Please try again.")
        end
    end
end

-- Initialize the program
io.write("Company Name: ")
local company_name = io.read()
local company_db = CompanyDB.new(company_name)
company_db:main_menu()
